### PR TITLE
line analysis: coerce units in results

### PR DESCRIPTION
**Pull request by kecnry**
### Description

This pull request converts the output quantities from line analysis so that it:
* combines all length units by converting all to the first found length unit with a power of 1
* strips any constants in the unit into the value

Before and after screenshots:
<div>
Screen Shot 2022-04-19 at 5 02 15 PM
Screen Shot 2022-04-19 at 5 00 44 PM
</div>

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [x] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [x] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Did the CI pass? If not, are the failures related?
- [x] Is a change log needed? If yes, is it added to `CHANGES.rst`?
- [x] Is a milestone set?
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

### CODE REVIEW

**Comment by rosteen** [comment]
When using this with the demo JWST spectrum (), I hit an error trace as soon as I open the `Line Analysis` plugin (see below), which works on main. Looks like it just needs to handle the case where the `uncertainty` attribute exists but is set to `None`.

```
AttributeError                            Traceback (most recent call last)
File ~/opt/anaconda3/envs/viz_dev/lib/python3.9/site-packages/ipywidgets/widgets/widget.py:676, in Widget._handle_msg(self, msg)
    674         if 'buffer_paths' in data:
    675             _put_buffers(state, data['buffer_paths'], msg['buffers'])
--> 676         self.set_state(state)
    678 # Handle a state request.
    679 elif method == 'request_state':

File ~/opt/anaconda3/envs/viz_dev/lib/python3.9/site-packages/ipywidgets/widgets/widget.py:545, in Widget.set_state(self, sync_data)
    542 if name in self.keys:
    543     from_json = self.trait_metadata(name, 'from_json',
    544                                     self._trait_from_json)
--> 545     self.set_trait(name, from_json(sync_data[name], self))

File ~/opt/anaconda3/envs/viz_dev/lib/python3.9/site-packages/glue_jupyter/state_traitlets_helpers.py:151, in GlueState.update_state_from_json(self, json, widget)
    149 self._block_on_state_change = True
    150 try:
--> 151     update_state_from_dict(state, json)
    152 finally:
    153     self._block_on_state_change = False

File ~/opt/anaconda3/envs/viz_dev/lib/python3.9/site-packages/glue_jupyter/state_traitlets_helpers.py:70, in update_state_from_dict(state, changes)
     68     setattr(state, name, get_cmap(changes[name]))
     69 else:
---> 70     setattr(state, name, changes[name])

File ~/opt/anaconda3/envs/viz_dev/lib/python3.9/site-packages/echo/core.py:261, in HasCallbackProperties.__setattr__(self, attribute, value)
    260 def __setattr__(self, attribute, value):
--> 261     super(HasCallbackProperties, self).__setattr__(attribute, value)
    262     if self.is_callback_property(attribute):
    263         self._notify_global(**{attribute: value})

File ~/opt/anaconda3/envs/viz_dev/lib/python3.9/site-packages/echo/core.py:76, in CallbackProperty.__set__(self, instance, value)
     74 new = self.__get__(instance)
     75 if old != new:
---> 76     self.notify(instance, old, new)

File ~/opt/anaconda3/envs/viz_dev/lib/python3.9/site-packages/echo/core.py:125, in CallbackProperty.notify(self, instance, old, new)
    123     return
    124 for cback in self._callbacks.get(instance, []):
--> 125     cback(new)
    126 for cback in self._2arg_callbacks.get(instance, []):
    127     cback(old, new)

File ~/projects/jdaviz/jdaviz/core/template_mixin.py:90, in PluginTemplateMixin._mxn_update_plugin_opened(self, new_value)
     88 app_state = self.app.state
     89 tray_names_open = [app_state.tray_items[i]['name'] for i in app_state.tray_items_open]
---> 90 self.plugin_opened = app_state.drawer and self._registry_name in tray_names_open

File ~/opt/anaconda3/envs/viz_dev/lib/python3.9/site-packages/traitlets/traitlets.py:606, in TraitType.__set__(self, obj, value)
    604     raise TraitError('The "%s" trait is read-only.' % self.name)
    605 else:
--> 606     self.set(obj, value)

File ~/opt/anaconda3/envs/viz_dev/lib/python3.9/site-packages/traitlets/traitlets.py:595, in TraitType.set(self, obj, value)
    591     silent = False
    592 if silent is not True:
    593     # we explicitly compare silent to True just in case the equality
    594     # comparison above returns something other than True/False
--> 595     obj._notify_trait(self.name, old_value, new_value)

File ~/opt/anaconda3/envs/viz_dev/lib/python3.9/site-packages/traitlets/traitlets.py:1219, in HasTraits._notify_trait(self, name, old_value, new_value)
   1218 def _notify_trait(self, name, old_value, new_value):
-> 1219     self.notify_change(Bunch(
   1220         name=name,
   1221         old=old_value,
   1222         new=new_value,
   1223         owner=self,
   1224         type='change',
   1225     ))

File ~/opt/anaconda3/envs/viz_dev/lib/python3.9/site-packages/ipywidgets/widgets/widget.py:606, in Widget.notify_change(self, change)
    603     if name in self.keys and self._should_send_property(name, getattr(self, name)):
    604         # Send new state to front-end
    605         self.send_state(key=name)
--> 606 super(Widget, self).notify_change(change)

File ~/opt/anaconda3/envs/viz_dev/lib/python3.9/site-packages/traitlets/traitlets.py:1229, in HasTraits.notify_change(self, change)
   1227 def notify_change(self, change):
   1228     """Notify observers of a change event"""
-> 1229     return self._notify_observers(change)

File ~/opt/anaconda3/envs/viz_dev/lib/python3.9/site-packages/traitlets/traitlets.py:1266, in HasTraits._notify_observers(self, event)
   1263 elif isinstance(c, EventHandler) and c.name is not None:
   1264     c = getattr(self, c.name)
-> 1266 c(event)

File ~/projects/jdaviz/jdaviz/configs/specviz/plugins/line_analysis/line_analysis.py:101, in LineAnalysis._on_plugin_opened_changed(self, *args)
     99     mark.visible = self.plugin_opened
    100 if self.plugin_opened:
--> 101     self._calculate_statistics()

File ~/projects/jdaviz/jdaviz/configs/specviz/plugins/line_analysis/line_analysis.py:321, in LineAnalysis._calculate_statistics(self, *args, **kwargs)
    318     else:
    319         temp_result = FUNCTIONS[function](spec_subtracted)
--> 321     temp_result = _coerce_unit(temp_result)
    322     temp_results.append({'function': function,
    323                          'result': str(temp_result.value),
    324                          'uncertainty': _uncertainty(temp_result),
    325                          'unit': str(temp_result.unit)})
    327 if not self.selected_line and self.identified_line:
    328     # default to the identified line

File ~/projects/jdaviz/jdaviz/configs/specviz/plugins/line_analysis/line_analysis.py:294, in LineAnalysis._calculate_statistics.<locals>._coerce_unit(quantity)
    292 coerced_quantity = quantity.to(coerced_unit_string)
    293 if hasattr(quantity, 'uncertainty'):
--> 294     coerced_quantity.uncertainty = quantity.uncertainty.to(coerced_unit_string)
    295 return coerced_quantity

AttributeError: 'NoneType' object has no attribute 'to'
```

File: jdaviz/configs/specviz/plugins/line_analysis/line_analysis.py [hunk 1]
```diff
@@ -271,6 +271,29 @@ def _uncertainty(result):
...
+                           in enumerate(zip(unit.bases, unit.powers, unit_types))
+                           if unit_type == 'length' and abs(power) == 1]
+            # we want to force all length units (not area) to use the same base unit so they can
+            # combine/cancel appropriately
+            coerced_bases = [unit.bases[i if i not in length_inds else length_inds[0]]
+                             for i in range(len(unit.bases))]
+            coerced_unit_string = ' * '.join([f'{base}**{power}'
+                                              for base, power in zip(coerced_bases, unit.powers)])
+            coerced_quantity = quantity.to(coerced_unit_string)
+            if hasattr(quantity, 'uncertainty'):
```
Comment (rosteen) [inline jdaviz/configs/specviz/plugins/line_analysis/line_analysis.py:293]:
```suggestion
            if hasattr(quantity, 'uncertainty') and quantity.uncertainty is not None:
```

Comment (kecnry) [reply jdaviz/configs/specviz/plugins/line_analysis/line_analysis.py]: agreed, good call!  I used `getattr(quantity, 'uncertainty', None)` just to be consistent with similar logic when parsing the uncertainty into the plugin results.

File: jdaviz/configs/specviz/plugins/line_analysis/line_analysis.py [hunk 2]
```diff
@@ -271,6 +271,29 @@ def _uncertainty(result):
             else:
                 return ''
 
+        def _coerce_unit(quantity):
+            """
+            coerce the unity on a quantity to have a single length unit (will take the first length
```
Comment (rosteen) [inline jdaviz/configs/specviz/plugins/line_analysis/line_analysis.py:276]:
```suggestion
            coerce the unit on a quantity to have a single length unit (will take the first length
```

**Review by rosteen — APPROVED** [review]
It feels bad that it can't be cleaner than this, but as far as I can tell it works. Made one minor suggestion to fix a typo, but other than that I think this is good to go.

**Review by javerbukh — APPROVED** [review]
Works as advertised.

**Comment by pllim** [comment]
I feel like maybe this could be simplified by using `Unit.compose()`, `Unit.decompose()`, `Unit.si`, or some built-in upstream machinery but I cannot be sure because I don't see any tests added here to declare the intent of this function.

**Comment by pllim** [comment]
p.s. Sorry for the late comment. I just got to this one in my pile.

**Comment by kecnry** [comment]
I tried using those and didn't have any luck, but if you have any suggestions, let me know!

Testing is a good point... this passes coverage because it hits those parts of the code, but there isn't a good test case there to make sure its doing what it should.  Should I split the function out of the class and add test cases to it alone?  Otherwise we need to test it through the plugin itself with data with these "odd" input units, and that might obfuscate the test logic.

**Comment by pllim** [comment]
👍 to unit test just for the backend function. Don't have to go crazy but do have to cover the units that the plugin cares about.

**Comment by pllim** [comment]
p.s. I might have an idea after I see the tests but no promises!

File: jdaviz/configs/specviz/plugins/line_analysis/tests/test_line_analysis.py [hunk 3]
```diff
@@ -80,3 +81,16 @@ def test_line_identify(specviz_helper, spectrum1d):
     la_plugin.selected_line = rest_names[1]
     assert ll_plugin.list_contents['Test List']['lines'][0].get('identify') is True
     assert ll_plugin.list_contents['Test List']['lines'][1].get('identify') is False
+
+
+def test_coerce_unit():
+    q_input = 1 * u.Unit('1E-20 erg m / (Angstrom cm**2 s)')
+    q_input.uncertainty = 0.1 * u.Unit('1E-20 erg m / (Angstrom cm**2 s)')
+    q_coerced = _coerce_unit(q_input)
+    assert q_coerced.unit == u.Unit('erg / (cm**2 s)')
```
Comment (pllim) [inline jdaviz/configs/specviz/plugins/line_analysis/tests/test_line_analysis.py:90]: What do you think?
```python
>>> q_input.unit.decompose(bases=[u.erg, u.cm, u.s])
Unit("1e-10 erg / (cm2 s)")
>>> q_input.decompose(bases=[u.erg, u.cm, u.s])
<Quantity 1.e-10 erg / (cm2 s)>
```

Comment (kecnry) [reply jdaviz/configs/specviz/plugins/line_analysis/tests/test_line_analysis.py:90]: that works for this case, but I think we want to support the _user_ bases (except for length, which will cancel once they become the same).

File: jdaviz/configs/specviz/plugins/line_analysis/tests/test_line_analysis.py [hunk 4]
```diff
@@ -80,3 +81,16 @@ def test_line_identify(specviz_helper, spectrum1d):
...
+    assert np.allclose(q_coerced.value, 1e-20 * u.m.to(u.Angstrom))
+    assert q_coerced.uncertainty.unit == u.Unit('erg / (cm**2 s)')
```
Comment (pllim) [inline jdaviz/configs/specviz/plugins/line_analysis/tests/test_line_analysis.py:92]:
```python
>>> q_input.uncertainty.decompose(bases=[u.erg, u.cm, u.s])
<Quantity 1.e-11 erg / (cm2 s)>
```

https://docs.astropy.org/en/latest/units/decomposing_and_composing.html#examples

File: jdaviz/configs/specviz/plugins/line_analysis/tests/test_line_analysis.py (hunk 3 above)
Comment (pllim) [reply jdaviz/configs/specviz/plugins/line_analysis/tests/test_line_analysis.py:90]: OK should we just merge then?

Comment (kecnry) [reply jdaviz/configs/specviz/plugins/line_analysis/tests/test_line_analysis.py:90]: Unless you think using decompose with the bases/powers logic I have would be cleaner than recompiling to a string?

Comment (pllim) [reply jdaviz/configs/specviz/plugins/line_analysis/tests/test_line_analysis.py:90]: Maybe not... as I have no idea what weird units will come with arbitrary data. Maybe we just ship this and revisit as needed.

**Comment by pllim** [comment]
Thanks!

**Pull request merged**